src/shogun/io/OpenMLFlow.h

```diff
@@ -0,0 +1,222 @@
+				std::string args_string = std::accumulate(
+				    args_vec.begin() + 1, args_vec.end(), args_vec.front(),
+				    [](std::string s0, std::string& s1) {
+					    return s0 += "/" + s1;
+				    });
+				request_path += args_string;
+			}
+
+			std::string url = request_format + request_path + "?" + m_api_key;
+
+			openml_curl_request_helper(url);
+
+			return m_curl_response_buffer;
+		}
+
+	private:
+		std::string m_curl_response_buffer;
+
+		/**
+		 * Initialises CURL session and gets the data.
+		 * This function also handles the response code from the server.
+		 *
+		 * @param url the url to query
+		 */
+		void openml_curl_request_helper(const std::string& url);
+
+		/**
+		 * Handles all possible codes
+		 *
+		 * @param curl_handle curl handle used in the request
+		 * @param code the code returned by the query
+		 */
+		void openml_curl_error_helper(CURL* curl_handle, CURLcode code);
+
+		std::string m_api_key;
+
+		static const char* xml_server;
+		static const char* json_server;
+
+		static const std::unordered_map<std::string, std::string>
+		    m_format_options;
+		static const std::unordered_map<std::string, std::string>
+		    m_request_options;
+
+		/* DATA API */
+		static const char* dataset_description;
+		static const char* list_data_qualities;
+		static const char* data_features;
+		static const char* list_dataset_qualities;
+		static const char* list_dataset_filter;
+
+		/* FLOW API */
+		static const char* flow_file;
+	};
+
+	class OpenMLWritter
+	{
+	public:
+		OpenMLWritter(const std::string& api_key) : m_api_key(api_key){};
+
+	private:
+		std::string m_api_key;
+	};
+
+	class OpenMLFlow
+	{
+
+	public:
+		using components_type =
+		std::unordered_map<std::string, std::shared_ptr<OpenMLFlow>>;
+		using parameters_type = std::unordered_map<
+				std::string, std::unordered_map<std::string, std::string>>;
+
+		OpenMLFlow(
+		    const std::string& name, const std::string& description,
+		    const std::string& model, components_type components,
+		    parameters_type parameters)
+		    : m_name(name), m_description(description), m_class_name(model),
+		      m_parameters(parameters), m_components(components)
+		{
+		}
+
+		~OpenMLFlow()= default;
+
+		static std::shared_ptr<OpenMLFlow>
+		download_flow(const std::string& flow_id, const std::string& api_key);
+
+		static void upload_flow(const std::shared_ptr<OpenMLFlow>& flow);
+
+		std::shared_ptr<OpenMLFlow> get_subflow(const std::string& name)
+		{
+			auto find_flow = m_components.find(name);
+			if (find_flow != m_components.end())
+				return find_flow->second;
+			else
+				SG_SERROR(
+				    "The provided subflow could not be found in this flow!")
+			return nullptr;
+		}
+
+	private:
+		std::string m_name;
+		std::string m_description;
+		std::string m_class_name;
+		parameters_type m_parameters;
+		components_type m_components;
+
+#ifndef SWIG
+		static void check_flow_response(rapidjson::Document& doc);
```

Comment:
does it really need to be part of the class, meaning can't you just have this function in the implementation?

Reply:
no, I guess I can move to free function and avoid exposing RapidJSON to swig :)